### PR TITLE
engine: resource: Add a Docker container resource

### PR DESCRIPTION
**Pull request by jonathangold**
I've put together the resource with what I feel is the minimum viable api. I still have to validate the input, but I wanted to put this up so we can discuss a few things in context.

Do we want to allow users to define stopped/terminated resources (like the aws resource?) 
Is there a better data structure than `map[string]string` for portmapping?

We need to talk about vendoring, because this involves vendoring a ~140mb repo, just to delete one of their vendored folders.

Feel free to read through the resource, but note, there aren't many comments at the moment.
Cheers.

### CODE REVIEW

File: engine/resources/docker.go [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+	// notify engine that we're running
+	if err := obj.init.Running(); err != nil {
+		return err // exit if requested
+	}
+
+	var send = false // send event?
+	for {
+		if obj.init.Debug {
+			obj.init.Logf("Watching: %s", passwdFile) // attempting to watch...
```
Comment (jonathangold) [inline engine/resources/docker.go]: Already changed. ;)

File: engine/resources/docker.go [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+)
+
+func init() {
+	engine.RegisterResource("docker", func() engine.Res { return &DockerRes{} })
+}
+
+// DockerRes is a docker account resource.
+type DockerRes struct {
```
Comment (purpleidea) [inline engine/resources/docker.go]: So since docker has turned into a behemoth, I'm wondering if it's best to do like aws and have various sub resources, like aws:ec2, etc... So we'd have docker:image, or whatever...

File: engine/resources/docker.go (hunk 1 above)
Comment (purpleidea) [reply engine/resources/docker.go]: you reviewed your own code, hehe, you can just push a new commit ;)

**Comment by purpleidea** [comment]
> jonathangold force-pushed the jonathangold:docker branch 5 times, most recently from 8efcc5d to 70710ef a day ago 

hehe... is GH getting more rude? ;)
review starting as requested...

File: engine/resources/docker.go [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/util"
+)
+
+const (
+	// dockerClientVersion is the api version used by the client. See:
+	// https://docs.docker.com/develop/sdk/#api-version-matrix for more info.
+	dockerClientVersion = "v1.37"
+	// dockerContainerRunning is the running container state.
+	containerRunning = "running"
```
Comment (purpleidea) [inline engine/resources/docker.go]: I think the states should probably be public in case we end up pulling them in from the lang somehow, but I think we'll eventually need a clean solution for these types of constants. This is sort of nitpicky, but I mentioned it to get you thinking about what we might eventually do on the language side.

File: engine/resources/docker.go (hunk 2 above)
Comment (jonathangold) [reply engine/resources/docker.go]: It makes sense to create new resources to implement different docker features. `docker:network` comes to mind. If we want to create a `docker:image` resource, it would be a matter of pulling out the image search and image pull requests (about six lines of code,) and plumbing in some auto-edges to the new resource. I think it makes sense to get this merged first with the image stuff, and pull it out later if we need to.

File: engine/resources/docker.go [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+	if obj.State != containerRunning && obj.State != containerStopped && obj.State != containerRemoved {
+		return fmt.Errorf("state must be running stopped or removed")
+	}
+
+	// validate image
+	c, err := client.NewClient(client.DefaultDockerHost, dockerClientVersion, nil, nil)
+	if err != nil {
+		return fmt.Errorf("error creating docker client: %s", err)
+	}
+	resp, err := c.ImageSearch(ctx, obj.Image, types.ImageSearchOptions{Limit: 1})
```
Comment (purpleidea) [inline engine/resources/docker.go]: You've brought up an interesting point here. My first thought was that these sorts of things should not happen in Validate... But maybe they should. Want to start a short discussion on the mailing list about why or why not we should run longer/network tasks in validate??

File: engine/resources/docker.go [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+	if err != nil {
+		return fmt.Errorf("error searching for image: %s", err)
+	}
+	if len(resp) == 0 {
+		return fmt.Errorf("image: %s not found", obj.Image)
+	}
+	c.Close()
```
Comment (purpleidea) [inline engine/resources/docker.go]: If we care about the return value (probably an err) then use it, otherwise this should be a defer in the right place I think.

File: engine/resources/docker.go [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+	// validate env
+	for _, env := range obj.Env {
+		if !strings.Contains(env, "=") || strings.Contains(env, " ") {
+			return fmt.Errorf("invalid environment variable: %s", env)
+		}
+	}
+
+	// validate ports
+	for k, v := range obj.Ports {
+		if k != "tcp" && k != "udp" {
```
Comment (purpleidea) [inline engine/resources/docker.go]: does docker support other protocols? sctp?

File: engine/resources/docker.go [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+	if err != nil {
+		return fmt.Errorf("error creating docker client: %s", err)
+	}
+
+	return nil
+}
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *DockerRes) Close() error {
+	obj.client.Close()
```
Comment (purpleidea) [inline engine/resources/docker.go]: just return this instead?

File: engine/resources/docker.go [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+	return nil
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *DockerRes) Watch() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt)
```
Comment (purpleidea) [inline engine/resources/docker.go]: why is this here?

File: engine/resources/docker.go [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+// CheckApply method for Docker resource.
+func (obj *DockerRes) CheckApply(apply bool) (checkOK bool, err error) {
+	obj.init.Logf("CheckApply(%t)", apply)
+
+	var id string
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt)
```
Comment (purpleidea) [inline engine/resources/docker.go]: see above

File: engine/resources/docker.go [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+			return false, fmt.Errorf("error creating container: %s", err)
+		}
+		id = resp.ID
+	}
+
+	if err := obj.client.ContainerStart(ctx, id, types.ContainerStartOptions{}); err != nil {
+		return false, fmt.Errorf("error starting container: %s", err)
+	}
+
+	// XXX: There is no ContainerWait for the running state, nor do we get any
```
Comment (purpleidea) [inline engine/resources/docker.go]: Can we add a brief spin-lock with a timeout then?

File: examples/lang/docker0.mcl [hunk 11]
```diff
@@ -0,0 +1,6 @@
+docker "mgmt-nginx" {
+    state => "running",
+    image => "nginx",
+    cmd => ["nginx", "-g", "daemon off;",],
```
Comment (purpleidea) [inline examples/lang/docker0.mcl]: out of curiosity, why is there a semicolon?

File: engine/resources/docker.go (hunk 2 above)
Comment (purpleidea) [reply engine/resources/docker.go]: Just saw this comment now. Sounds good. Shouldn't we name this something different? docker:???

File: examples/lang/docker0.mcl (hunk 11 above)
Comment (jonathangold) [reply examples/lang/docker0.mcl]: Copied directly from the nginx dockerfile. :P
https://github.com/nginxinc/docker-nginx/blob/f4d30145c60c433966df96f618d78513fee9d322/mainline/stretch/Dockerfile#L99

File: engine/resources/docker.go (hunk 8 above)
Comment (jonathangold) [reply engine/resources/docker.go]: To cancel the context in the event that we want to exit, and there are api requests still outstanding. Should I use a struct context and cancel it in close instead?

File: engine/resources/docker.go (hunk 6 above)
Comment (jonathangold) [reply engine/resources/docker.go]: Yup. Missed sctp. Good catch. :)

File: engine/resources/docker.go (hunk 2 above)
Comment (jonathangold) [reply engine/resources/docker.go]: Yeah. `docker:container` is boring but descriptive. Any better ideas? Also, do colons work with the language yet?

Comment (purpleidea) [reply engine/resources/docker.go]: Yes in theory, but not in git master yet, and not tested yet:

https://github.com/purpleidea/mgmt/blob/6405b2371bb7a03a23ab67890fc4530930527f7f/lang/lexer.nex#L362

You can try branch feat/import for new things, but might be broken, eat your dog, get rebased often, etc...

File: engine/resources/docker.go [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+	// ContainerRunning is the running container state.
+	ContainerRunning = "running"
+	// ContainerStopped is the stopped container state.
+	ContainerStopped = "stopped"
+	// ContainerRemoved is the removed container state.
+	ContainerRemoved = "removed"
+
+	// dockerClientVersion is the api version used by the client. See:
+	// https://docs.docker.com/develop/sdk/#api-version-matrix for more info.
+	dockerClientVersion = "v1.37"
```
Comment (purpleidea) [inline engine/resources/docker.go]: So it bugged me that we'd have some constant in here which would potentially rot over time, so I did some source diving, and: api.DefaultVersion exists! woo. But even better, it seems if we use the empty string, then that is taken as the default. Woo, woo!

So I think you should move this to a struct arg that defaults to empty string. Unless it's never something you want to specify in which case just empty string. Also there's this: https://github.com/moby/moby/blob/e6244aa598e9894ead8445ca66329d26eb016e64/client/client.go#L249

What do you think? I'm leaning on struct option + warning to specify a version if desired.

File: engine/resources/docker.go [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+	eventOpts := types.EventsOptions{
+		Filters: filters.NewArgs(filters.KeyValuePair{"container", id}),
+	}
+	eventCh, errCh := client.Events(ctx, eventOpts)
+	// Start the container.
+	if err := client.ContainerStart(ctx, id, opts); err != nil {
+		return fmt.Errorf("error starting container: %s", err)
+	}
+	// Wait for a message on eventChan that says the container has started.
+	select {
```
Comment (purpleidea) [inline engine/resources/docker.go]: Is it possible that either of these never occurs and we block indefinitely?

Comment (jonathangold) [reply engine/resources/docker.go]: I don't think so but I'll add a `case <- ctx.Done():` for safety.

File: engine/resources/docker.go (hunk 12 above)
Comment (jonathangold) [reply engine/resources/docker.go]: My only concern is that the way certain methods work, or if they're even available at all, may change depending on the API version, leading to undefined behaviour. I think `api.DefaultVersion` is probably the safest bet, and fits with your "do it with git master" philosophy.

Comment (jonathangold) [reply engine/resources/docker.go]: :(
```
18:43:23 main.go:186: main: error validating the new graph: error searching for image: Error response from daemon: client version 1.40 is too new. Maximum supported API version is 1.39
```

**Comment by purpleidea** [comment]
Ready for a review?

File: engine/resources/docker_container.go [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+func init() {
+	engine.RegisterResource("docker:container", func() engine.Res { return &DockerContainerRes{} })
+}
+
+// DockerContainerRes is a docker container resource.
+type DockerContainerRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Edgeable
+
+	State      string                     `yaml:"state"`      // running, stopped, or removed
```
Comment (purpleidea) [inline engine/resources/docker_container.go]: Put these lovely comments in the correct format above each struct field. They'll become docs!

File: engine/resources/docker_container.go [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+	// validate state
+	if obj.State != ContainerRunning && obj.State != ContainerStopped && obj.State != ContainerRemoved {
+		return fmt.Errorf("state must be running, stopped or removed")
+	}
+
+	// TODO: should this stuff happen in validate or do we make sure there's no
+	// whitespace, and let it fail in CheckApply?
+	//
+	// set up temporary docker client to validate image
+	c, err := client.NewClient(client.DefaultDockerHost, obj.APIVersion, nil, nil)
```
Comment (purpleidea) [inline engine/resources/docker_container.go]: Let's not make any connections in Validate()
Keep it "cheap" so we can casually use Validate() in many places.
Add these checks to Init() or CheckApply. Init seems more likely.
Of course, anything you can do without the connection should be done in Validate.
Why? Well we might consider adding Validate() calls from the lang in order to statically check things that are determined statically as a compiler optimization. Eg: If you asked for state => "wrong", since those are all constants we can catch this at compile time... Only if Validate is cheap =D

File: engine/resources/docker_container.go [hunk 16]
```diff
@@ -0,0 +1,441 @@
...
+		}
+	}
+
+	// validate ports
+	for k, v := range obj.Ports {
+		if k != "tcp" && k != "udp" && k != "sctp" {
+			return fmt.Errorf("ports primary key should be tcp, udp or sctp")
+		}
+		for p, q := range v {
+			if (p < 1 || p > 0xFFFF) || (q < 1 || q > 0xFFFF) {
```
Comment (purpleidea) [inline engine/resources/docker_container.go]: why the 0x instead of just an int?

File: engine/resources/docker_container.go [hunk 17]
```diff
@@ -0,0 +1,441 @@
...
+				return fmt.Errorf("ports should be between 1 and 65535")
```
Comment (purpleidea) [inline engine/resources/docker_container.go]: s/should be/must be/ right?

File: engine/resources/docker_container.go [hunk 18]
```diff
@@ -0,0 +1,441 @@
...
+			}
+		}
+	}
+
+	// validate APIVersion
+	if obj.APIVersion != "" {
+		verOK, err := regexp.MatchString(`^(v)[1-9]\.[0-9]\d*$`, obj.APIVersion)
+		if err != nil {
+			return fmt.Errorf("error matching apiversion string: %s", err)
```
Comment (purpleidea) [inline engine/resources/docker_container.go]: errwrap

File: engine/resources/docker_container.go [hunk 19]
```diff
@@ -0,0 +1,441 @@
...
+		}
+
+		c, err := obj.client.ContainerCreate(ctx, containerConfig, hostConfig, nil, obj.Name())
+		if err != nil {
+			return false, fmt.Errorf("error creating container: %s", err)
+		}
+		id = c.ID
+	}
+
+	return false, containerStart(ctx, obj.client, id, types.ContainerStartOptions{})
```
Comment (purpleidea) [inline engine/resources/docker_container.go]: why do you pass around obj.client instead of just having everything be a private struct method and using it directly??

File: lang/lexer.nex [hunk 20]
```diff
@@ -332,7 +332,7 @@
 			lval.str = strings.ToLower(s) // uncapitalize it
 			return CAPITALIZED_IDENTIFIER
 		}
-/[a-z][a-z0-9]*/
+/[a-z]([a-z0-9:]*[a-z0-9]+)?/
```
Comment (purpleidea) [inline lang/lexer.nex]: Unfortunately please wait for this to land in git master from my branch...
It means you'll just have to change it manually to test the resource, sorry.

File: engine/resources/docker_container.go (hunk 16 above)
Comment (jonathangold) [reply engine/resources/docker_container.go]: I think 0xFFFF better communicates the reason for the check (max value of uint16). Seems less arbitrary than 65536. I wish go had a simple build-in max and min-value methods on number types. How would you prefer I do it, int const?

File: engine/resources/docker_container.go (hunk 19 above)
Comment (jonathangold) [reply engine/resources/docker_container.go]: It made the tests a little easier to write, but I don't mind changing it if you don't like it.

File: engine/resources/docker_container.go (hunk 16 above)
Comment (purpleidea) [reply engine/resources/docker_container.go]: yeah, just put the raw int there. 65535 I think, right? not 65536 ??

File: engine/resources/docker_container.go (hunk 19 above)
Comment (purpleidea) [reply engine/resources/docker_container.go]: Ah. Interesting point. I think you can just init the tests with &DockerContainerRes{client: whatever,} and then call the methods like that.

Comment (jonathangold) [reply engine/resources/docker_container.go]: Yup, will do.

**Comment by jonathangold** [comment]
@purpleidea I deleted my fork when I was playing with hub, so I have to open a new PR, since it can't find my branch.
